cpp/src/arrow/io/memory.h

```diff
@@ -38,11 +42,49 @@ class Status;
 
 namespace io {
 
+class ARROW_EXPORT Memcopy {
+public:
+  virtual void memcopy(uint8_t* dst, const uint8_t* src, uint64_t nbytes) = 0;
```

Comment:
Yeah, I will make these changes in a bit, right now I'm still figuring out the benchmark ;)